rasa/core/actions/action.py

```diff
@@ -147,41 +125,53 @@ def is_retrieval_action(action_name: Text, retrieval_intents: List[Text]) -> boo
     )
 
 
-def action_from_name(
-    name: Text, domain: Domain, action_endpoint: Optional[EndpointConfig]
+def action_for_name_or_text(
+    action_name_or_text: Text, domain: Domain, action_endpoint: Optional[EndpointConfig]
 ) -> "Action":
     """Retrieves an action by its name.
 
     Args:
-        name: The name of the action.
+        action_name_or_text: The name of the action.
         domain: The current model domain.
         action_endpoint: The endpoint to execute custom actions.
 
+    Raises:
+        ActionNotFoundException: If action not in current domain.
     Returns:
         The instantiated action.
     """
+
+    if action_name_or_text not in domain.action_names:
+        domain.raise_action_not_found_exception(action_name_or_text)
+
     defaults = {a.name(): a for a in default_actions(action_endpoint)}
 
-    if name in defaults and name not in domain.user_actions_and_forms:
-        return defaults[name]
+    if (
+        action_name_or_text in defaults
+        and action_name_or_text not in domain.user_actions_and_forms
+    ):
+        return defaults[action_name_or_text]
 
-    if name.startswith(UTTER_PREFIX) and is_retrieval_action(
-        name, domain.retrieval_intents
+    if action_name_or_text.startswith(UTTER_PREFIX) and is_retrieval_action(
+        action_name_or_text, domain.retrieval_intents
     ):
-        return ActionRetrieveResponse(name)
+        return ActionRetrieveResponse(action_name_or_text)
+
+    if action_name_or_text in domain.action_texts:
+        return EndToEndAction(action_name_or_text)
```

Comment:
this is the essential part 1 of the change

Reply:
I'd say logically it should go before checking `.startswith(UTTER_PREFIX)` on Line 155

Reply:
Mhm, one could argue that having it after it is better for performance as the `utter_` case should be way more common.

Reply:
but what if your action_text is "utter_goodbye should be called when user says goodbye"?